### PR TITLE
Enable setting build-dir

### PR DESCRIPTION
**Pull request by nanonyme**
build-dir is disabled by default and can be enabled with project conf
or through element

### CODE REVIEW

**Comment by nanonyme** [comment]
Continuation of https://github.com/apache/buildstream-plugins/pull/21

File: src/buildstream_plugins/elements/autotools.yaml [hunk 1]
```diff
@@ -93,20 +94,30 @@ config:
   - |
     %{autogen}
   - |
-    %{configure}
+    if [ -n "%{build-dir}" ]; then
+      mkdir -p "%{build-dir}"
+      cd "%{build-dir}"
+    fi
+    %{configure})
```
Comment (gtristan) [inline src/buildstream_plugins/elements/autotools.yaml:101]: Trailing `)` ?

File: src/buildstream_plugins/elements/autotools.yaml [hunk 2]
```diff
@@ -93,20 +94,30 @@ config:
...
   # Commands for installing the software into a
   # destination folder
   #
   install-commands:
   - |
-    %{make-install}
+    if [ -n "%{build-dir}" ]; then
+      cd "%{build-dir}"
+    fi
+    %{make-install})
```
Comment (gtristan) [inline src/buildstream_plugins/elements/autotools.yaml:120]: Trailing `)` ?

**Comment by gtristan** [comment]
I think this would be more flexible and convenient if:
* the default for `build-dir` was `.`
* the `make` and `make-install` variables were extended to contain `-C %{build-dir}`

it may also be partly a matter of taste, I think this would be much more elegant than having the script jumping around into different directories.

**Comment by nanonyme** [comment]
Jumping still required for configure.

File: src/buildstream_plugins/elements/autotools.py [hunk 3]
```diff
@@ -51,6 +51,14 @@
          conf-global: |
            --disable-gtk-doc --disable-static
 
+By default build-dir is not used. To improve probability build does not
+tamper with sources, you can set build-dir like
+
+.. code:: yaml
+
+   variables:
+     build-dir: bst_build-dir
```
Comment (gtristan) [inline src/buildstream_plugins/elements/autotools.py:60]: Nit: this is a strange suggestion with mixed `-` and `_`

perhaps `_build` ?

Comment (nanonyme) [reply src/buildstream_plugins/elements/autotools.py:60]: Ah, yes. True.

**Comment by gtristan** [comment]
Overall looks good to me

File: src/buildstream_plugins/elements/autotools.py (hunk 3 above)
Comment (nanonyme) [reply src/buildstream_plugins/elements/autotools.py:60]: I guess to be consistent with cmake and meson we could suggest _builddir

**Comment by nanonyme** [comment]
@gtristan this should now be ready. Independent of this all cache keys are changed with bst master.

File: src/buildstream_plugins/elements/autotools.yaml [hunk 4]
```diff
@@ -93,6 +94,8 @@ config:
   - |
     %{autogen}
   - |
+    mkdir -p "%{build-dir}"
+    cd "%{build-dir}"
     %{configure}
```
Comment (abderrahim) [inline src/buildstream_plugins/elements/autotools.yaml:99]: Don't we need a `cd -` here?

**Review by abderrahim — APPROVED** [review]
Other than that, LGTM.

File: src/buildstream_plugins/elements/autotools.yaml (hunk 4 above)
Comment (nanonyme) [reply src/buildstream_plugins/elements/autotools.yaml:99]: Turns out we don't. These aren't actually concatenated into one shell script.

Comment (nanonyme) [reply src/buildstream_plugins/elements/autotools.yaml:99]: @abderrahim did I misunderstand what you were after?

Comment (gtristan) [reply src/buildstream_plugins/elements/autotools.yaml:99]: Indeed we do not need `cd -`.

**Comment by gtristan** [comment]
> @gtristan this should now be ready. Independent of this all cache keys are changed with bst master.

I've updated `buildstream-plugins` and it's cache key test.

Can you please rebase this branch ?

**Comment by nanonyme** [comment]
@gtristan rebase is now done.

**Pull request merged**